viofs/svc/virtiofs.c

```diff
@@ -135,6 +163,197 @@ static void FUSE_HEADER_INIT(struct fuse_in_header *hdr, uint32_t opcode,
+            Notification, NULL);
+    if (Notification->UnregWork == NULL)
+    {
+        return FALSE;
+    }
+
+    return TRUE;
+}
+
+static VOID VirtFsNotificationDelete(VIRTFS_NOTIFICATION *Notification)
+{
+    CloseThreadpoolWork(Notification->UnregWork);
+    DeleteCriticalSection(&Notification->Lock);
+}
+
+static DWORD VirtFsRegDevHandleNotification(VIRTFS *VirtFs)
+{
+    HCMNOTIFICATION Handle = NULL;
+    CM_NOTIFY_FILTER Filter;
+    CONFIGRET ConfigRet;
+
+    ZeroMemory(&Filter, sizeof(Filter));
+    Filter.cbSize = sizeof(Filter);
+    Filter.FilterType = CM_NOTIFY_FILTER_TYPE_DEVICEHANDLE;
+    Filter.u.DeviceHandle.hTarget = VirtFs->Device;
+
+    ConfigRet = CM_Register_Notification(&Filter, VirtFs,
+            DeviceNotificationCallback, &Handle);
+
+    if (ConfigRet == CR_SUCCESS)
+    {
+        VirtFs->DevHandleNotification.Handle = Handle;
+        VirtFs->DevHandleNotification.Unregister = FALSE;
+    }
+
+    return CM_MapCrToWin32Err(ConfigRet, ERROR_NOT_SUPPORTED);
+}
+
+static DWORD VirtFsDevInterfaceArrival(VIRTFS *VirtFs, HCMNOTIFICATION Notify)
+{
+    DWORD Error;
+    NTSTATUS Status;
+    VIRTFS_NOTIFICATION *Notification = &VirtFs->DevHandleNotification;
+
+    DBG("Notify = 0x%x", Notify);
+
+    // Wait for unregister work to end, if any.
+    WaitForThreadpoolWorkCallbacks(Notification->UnregWork, FALSE);
+
+    Error = FindDeviceInterface(&VirtFs->Device);
+    if (Error != ERROR_SUCCESS)
+    {
+        return Error;
+    }
+
+    Error = VirtFsRegDevHandleNotification(VirtFs);
+    if (Error != ERROR_SUCCESS)
+    {
+        goto out_close_handle;
+    }
+
+    Status = VirtFsStart(VirtFs);
+    if (!NT_SUCCESS(Status))
+    {
+        Error = FspWin32FromNtStatus(Status);
+        goto out_unreg_dh_notify;
+    }
+
+    return ERROR_SUCCESS;
+
+out_unreg_dh_notify:
+    VirtFsNotificationAsyncUnreg(&VirtFs->DevHandleNotification);
+out_close_handle:
+    CloseHandle(VirtFs->Device);
+
+    return Error;
+}
+
+static VOID VirtFsDevQueryRemove(VIRTFS *VirtFs, HCMNOTIFICATION Notify)
+{
+    DBG("Notify = 0x%x", Notify);
+
+    FspFileSystemStopDispatcher(VirtFs->FileSystem);
+    VirtFsDelete(VirtFs);
+    VirtFsNotificationAsyncUnreg(&VirtFs->DevHandleNotification);
+}
+
+DWORD WINAPI DeviceNotificationCallback(HCMNOTIFICATION Notify,
+    PVOID Context, CM_NOTIFY_ACTION Action, PCM_NOTIFY_EVENT_DATA EventData,
+    DWORD EventDataSize)
+{
+    VIRTFS *VirtFs = Context;
+    DWORD Result = ERROR_SUCCESS;
+
+    UNREFERENCED_PARAMETER(EventData);
+    UNREFERENCED_PARAMETER(EventDataSize);
+
+    switch (Action)
+    {
+        case CM_NOTIFY_ACTION_DEVICEINTERFACEARRIVAL:
+        case CM_NOTIFY_ACTION_DEVICEQUERYREMOVEFAILED:
+            Result = VirtFsDevInterfaceArrival(VirtFs, Notify);
+            break;
+        case CM_NOTIFY_ACTION_DEVICEQUERYREMOVE:
+        case CM_NOTIFY_ACTION_DEVICEREMOVECOMPLETE:
+            VirtFsDevQueryRemove(VirtFs, Notify);
```

Comment:
Is VirtFsDevQueryRemove fully protected from repetitive calls?

Reply:
It wasn't, but I've fixed it.